### DISCUSSION TITLE
a webpack-api library should be added ? 🤔

### DISCUSSION THREAD

**Discussion by xiaoxiaojx** [webpack/webpack#13934] [Ideas]
👋  hi, all 
> This idea comes from the [napi](https://github.com/nodejs/node/blob/master/src/node_api.cc) and [node-addon-api](https://github.com/nodejs/node-addon-api) implementations

In the plugins I've written, I've often written them in the following way
```
if ("hooks" in compiler) {
      // webpack 4
      eventName = trainCaseToCamelCase(eventName);
      if (compiler.hooks[eventName]) {
          compiler.hooks[eventName][isAsync ? "tapAsync" : "tap"](
              pluginName,
              func
          );
      }
  } else {
      // webpack 2/3
      compiler.plugin(eventName, func);
  }
```
or implement a dynamically increasing entry like nextjs
```
// webpack 4

const dep = MultiEntryPlugin.createDependency(entry, name)
compilation.addEntry(context, dep, name, err => {
    if (err) return reject(err)
    resolve()
})
```
```
// webpack 5

const EntryOptionPlugin = require('webpack/lib/EntryOptionPlugin')
EntryOptionPlugin.entryDescriptionToOptions(compiler, name, {
    import: entry
})
```
Having a stable api library would reduce the cost for plugin developers and make it easy for them to write plugins that are compatible with all versions.

What about this 💡  idea ? 🤔
> If feasible, I hope to join this interesting project ...

**Comment by alexander-akait** [comment]
No need, we don't have a lot of major release and keep API as long as possible, it is normal when major releases have breaking changes

**Reply by alexander-akait** [reply]
And time to drop webpack v2/v3 and v4 support, official plugins and loaders already done this (most of them)

**Reply by xiaoxiaojx** [reply]
My heart is broken 💔

**Reply by alexander-akait** [reply]
Create another package requires a lot of time for support and compatibility, unfortunately we do not have such capabilities

**Reply by xiaoxiaojx** [reply]
This needs to collect the high-frequency problems currently used, or other plugin developers to integrate the compatible processing solutions written here ......
